src/test/scala/io/iohk/ethereum/blockchain/sync/fast/FastSyncBranchResolverSpec.scala

```diff
@@ -0,0 +1,309 @@
+package io.iohk.ethereum.blockchain.sync.fast
+
+import org.scalatest.wordspec.AnyWordSpec
+import org.scalatest.matchers.must.Matchers
+import org.scalamock.scalatest.MockFactory
+import io.iohk.ethereum.domain.Blockchain
+import io.iohk.ethereum.domain.BlockHeader
+import org.scalatestplus.scalacheck.ScalaCheckPropertyChecks
+import org.scalacheck.Gen
+import io.iohk.ethereum.domain.Block
+import akka.util.ByteString
+import io.iohk.ethereum.domain.BlockchainImpl
+import io.iohk.ethereum.BlockHelpers
+import io.iohk.ethereum.blockchain.sync.fast.FastSyncBranchResolver.SearchState
+import io.iohk.ethereum.network.Peer
+import java.net.InetSocketAddress
+import akka.actor.ActorRef
+import io.iohk.ethereum.blockchain.sync.fast.BinarySearchSupport._
+import io.iohk.ethereum.Fixtures
+
+class FastSyncBranchResolverSpec extends AnyWordSpec with Matchers with MockFactory {
+
+  import Fixtures.Blocks.ValidBlock
+
+  private def blocksMap(amount: Int, parent: Block): Map[BigInt, Block] = {
+    BlockHelpers.generateChain(amount, parent).map(b => (b.number, b)).toMap
+  }
+
+  private def headersMap(amount: Int, parent: Block): Map[BigInt, BlockHeader] = {
+    BlockHelpers.generateChain(amount, parent).map(b => (b.number, b.header)).toMap
+  }
+
+  private def headersList(blocksMap: Map[BigInt, Block]): List[BlockHeader] =
+    blocksMap.values.map(_.header).toList
+
+  private def headerNum(block: Block): BigInt = block.header.number
+
+  "FastSyncBranchResolver" must {
+    "calculate childOf block number" in {
+      import FastSyncBranchResolver.childOf
+      assert(childOf(0) === 1)
+      assert(childOf(6) === 7)
+    }
+    "calculate parentOf block number" in {
+      import FastSyncBranchResolver.parentOf
+      assert(parentOf(7) === 6)
+    }
+    "correctly discard all blocks after a certain block number" in {
+      val mockedBlockchain = mock[BlockchainImpl]
+
+      val headers = headersMap(amount = 3, parent = Block(ValidBlock.header.copy(number = 97), ValidBlock.body))
+
+      inSequence {
+        (mockedBlockchain.getBestBlockNumber _).expects().returning(BigInt(100)).once()
+        (mockedBlockchain.getBlockHeaderByNumber _).expects(BigInt(100)).returning(headers.get(100))
+        (mockedBlockchain.removeBlock _).expects(headers(100).hash, false).returning(())
+        (mockedBlockchain.getBlockHeaderByNumber _).expects(BigInt(99)).returning(headers.get(99))
+        (mockedBlockchain.removeBlock _).expects(headers(99).hash, false).returning(())
+        (mockedBlockchain.getBlockHeaderByNumber _).expects(BigInt(98)).returning(headers.get(98))
+        (mockedBlockchain.removeBlock _).expects(headers(98).hash, false).returning(())
+      }
+
+      val resolver = new FastSyncBranchResolver {
+        override val blockchain: Blockchain = mockedBlockchain
+      }
+      resolver.discardBlocksAfter(97)
```

Comment:
This test seems to miss a final assertion

Reply:
No, it does not...it's testing the expected calls to `Blockchain`. Whether `Blockchain` removes blocks correctly is not the responsibility of the branch resolver.

Reply:
👍